### ISSUE TITLE
Postgresql COPY command

### ISSUE THREAD

**Issue by ClasicRando** [jasync-sql/jasync-sql#419]
Are there any plans to include support for the COPY command within the postgres driver? If not, would there be interest in accepting any community contributions to make the feature available?

**Comment by oshai** [comment]
I don't have such plans but contribution is welcome.

**Comment by ClasicRando** [comment]
@oshai  I have created my initial attempt with some tests created to check the new behaviour. I'll continue to work on tests to make sure I don't break anything.

This will be a larger than excepted change since the connection technically changes state during a copy operation (i.e. I need to disallow querying while either CopyIn or CopyOut is active). I also had a very hard time allowing concurrent reading of CopyOut data because of race conditions that arise if I allow the user to attempt reading each row with a future based return. I opted for an iterator based approach to allow for concurrent reading of postgres stream data into a queue for the iterator to pull from. However, this would technically introduce locking (through a BlockingQueue) so any advice on another approach would be appreciated if you do not want any locking on the client side.

CopyIn is pretty straight forward since the user supplies chunks of bytes to the connection to pass forward to the server so that feature is essentially solidified.

**Comment by oshai** [comment]
> I need to disallow querying while either CopyIn or CopyOut is active.

I don't think you need to add support for it. You can see [here](https://github.com/jasync-sql/jasync-sql/blob/83029e5d8c7405b87c15376383afe5d8fecf9e06/postgresql-async/src/main/java/com/github/jasync/sql/db/postgresql/PostgreSQLConnection.kt#L386) how we prevent concurrent queries.

Regarding reading in parallel, in regular queries we collect all result before the future is completed. Maybe do something similar here as well? Not sure I got the issue right, as netty promise us we will not have multiple reads from the socket in parallel.

**Comment by ClasicRando** [comment]
For either copy operation the concurrent queries check could help, however, issuing a copy query to the server has a different flow than regular queries.
For CopyOut it goes:
1. `COPY TO STDOUT` query sent
2. Server responds with CopyOutResponse [docs](https://www.postgresql.org/docs/current/protocol-message-formats.html#PROTOCOL-MESSAGE-FORMATS-COPYOUTRESPONSE)
3. The server then sends zero or more CopyData messages ([docs](https://www.postgresql.org/docs/current/protocol-message-formats.html#PROTOCOL-MESSAGE-FORMATS-COPYDATA)) where each message represents a row as a single collection of bytes
4. Once all the rows have been sent, the server finalizes the copy with a CopyDone message [docs](https://www.postgresql.org/docs/current/protocol-message-formats.html#PROTOCOL-MESSAGE-FORMATS-COPYDONE)
5. With all rows of the target table copied to the client, the server send the customary CommandComplete followed by the ReadyForQuery messages.

So for CopyOut, if we want to push each row to the query result, we can hook into the existing query processing with special cases for the CopyOut command. My only concern here would be holding a large amount of data in memory since the CopyOut command is usually used for bulk data processing. An iterator might help if the client is able to process incoming rows fairly quickly and limit the memory consumption for bulk operations, but again, that would pose it's own issues.

For CopyIn:
1. `COPY FROM STDIN` query sent
2. Server response with CopyInResponse [docs](https://www.postgresql.org/docs/current/protocol-message-formats.html#PROTOCOL-MESSAGE-FORMATS-COPYINRESPONSE)
3. This now lets the user send zero or more rows to the server as CopyData messages (see docs above)
4. Once the user is done pushing data to the server, the user must send a CopyDone mesage (see docs above)
5. The server then responds with the customary CommandComplete followed by the ReadyForQuery messages.

So for CopyIn, since the client is in control of completing the copy operation, the client might accidentally send a regular query message while a copy in is active which (to my understanding) will disrupt the message flow and lead to a connection state I cannot account for.

All of this is to say that the Copy command presents some challenges to how the event handling works with postgres connections. What I could do is set the query promise to a non empty value for the duration of the copy operation to avoid querying the server while in copy mode. This also might be made a little easier if I was to create dedicated `copyIn`/`copyOut` methods on the `PostgreSQLConnection` class since currently, the copy messages are treated as raw queries.

**Comment by oshai** [comment]
Thanks for all the details!
One question first - what is the benefit of using copy over insert / select? is it just convenience?

> we can hook into the existing query processing with special cases for the CopyOut command.

As you mentioned the safest way is to gather all data beforehand (that's what we do with regular queries). An iterator sounds like an interesting approach (maybe a flux?). Currently the lib is async but not reactive, so do not support those constructs.

>So for CopyIn, since the client is in control of completing the copy operation, the client might accidentally send a regular query message while a copy in is active which (to my understanding) will disrupt the message flow and lead to a connection state I cannot account for.

If you set the query promise this will not be allowed.

> This also might be made a little easier if I was to create dedicated copyIn/copyOut methods on the PostgreSQLConnection class since currently, the copy messages are treated as raw queries.

We can start like that if it's more convenient and decide later.

**Comment by erikmata** [comment]
With regard to the question about the difference between copy and regular inserts in Postgresql:

Having the ability to issue and handle COPY commands would be very useful indeed.

Using copy can be orders of magnitude more efficient than regular inserts. Other database systems (MS SQL, Oracle) also provide such functionality: sometimes called bulk import (not the same as batch import). It is a way to short-circuit pumping in large volumes of data into the database. Copy can be issued from STDIN, from FILE or from PROGRAM.

My humble understanding is that, with Postgresql (and probably also MS SQL/Oracle), the downside of COPY is that you as a user must make sure that the data conforms to the constraints of the target table. If the target table contains unique keys (e.g. primary key), the user must make sure that the source data doesn't contain any duplicate values for those unique keys. Furthermore, if the target table has constraints on allowed values for certain columns, then the user must make sure that the source data only contains such allowed values. Any issues during copy, i.e. any row in the source data that doesn't conform to the constraints of the target table, will lead to Postgresql issuing an error and the whole transaction (copy command) will be rolled back.

Certain use cases (e.g event sourcing) are happy without unique keys or other constraints.

For other use cases that need constraints, this downside can be mitigated. One strategy that can be used to avoid failures all together, is to create a temporary target table that doesn't have any constraints at all, and copy the data into the temporary table.
Then, in a second step, one can issue a regular insert from the temporary table to the final target table (that has constraints). In this regular insert, it's possible to handle conflicts in unique keys (using upsert/ignore), as well as, transform disallowed values into allowed values.

You might wonder what the upside is to issue a COPY followed by an INSERT? My experience is that, with this strategy, the source data can be transferred via efficient file operations from the client to the server system, then imported into the database using a single COPY followed by a single INSERT. Heavy processing is done on the server side, the client just issues two stateless commands.

On the other hand, using only regular INSERTs, if the source data volume is large, the client must issue multiple INSERT commands, folding chunks of processed source data into those commands.

Since I discovered the COPY command in Postgresql, I have completely stopped using client tools (such as DBeaver, DataGrip) to export/import data from/to my database. With COPY I can rest assured that the exported CSV/binary format is compatible with the corresponding import format.
